src/base/QXmppHttpUploadIq.cpp

```
@@ -0,0 +1,253 @@
+    writer->writeAttribute("filename", d->fileName);
+    writer->writeAttribute("size", QString::number(d->size));
+    // content-type is optional
+    if (!d->contentType.isDefault() && d->contentType.isValid())
+        writer->writeAttribute("content-type", d->contentType.name());
+    writer->writeEndElement();
+}
+/// \endcond
+
+class QXmppHttpUploadSlotIqPrivate
+{
+public:
+    QUrl putUrl;
+    QUrl getUrl;
+    QMap<QString, QString> putHeaders;
+};
+
+QXmppHttpUploadSlotIq::QXmppHttpUploadSlotIq()
+    : d(new QXmppHttpUploadSlotIqPrivate())
+{
+}
+
+QXmppHttpUploadSlotIq::~QXmppHttpUploadSlotIq()
+{
+    delete d;
+}
+
+/// Returns the URL for uploading via. HTTP PUT.
+
+QUrl QXmppHttpUploadSlotIq::putUrl() const
+{
+    return d->putUrl;
+}
+
+/// Sets the URL the client should use for uploading.
+
+void QXmppHttpUploadSlotIq::setPutUrl(const QUrl &putUrl)
+{
+    d->putUrl = putUrl;
+}
+
+/// Returns the URL to where the file will be served.
+
+QUrl QXmppHttpUploadSlotIq::getUrl() const
+{
+    return d->getUrl;
+}
+
+/// Sets the download URL.
+
+void QXmppHttpUploadSlotIq::setGetUrl(const QUrl &getUrl)
+{
+    d->getUrl = getUrl;
+}
+
+/// Returns a map of header fields (header name -> value) that need to be
+/// included in the PUT (upload) request. This won't contain any other fields
+/// than: "Authorization", "Cookie" or "Expires".
+
+QMap<QString, QString> QXmppHttpUploadSlotIq::putHeaders() const
+{
+    return d->putHeaders;
+}
+
+/// Sets the header fields the client needs to include in the PUT (upload)
+/// request. All fields other than "Authorization", "Cookie" or "Expires" will
+/// be ignored.
+
+void QXmppHttpUploadSlotIq::setPutHeaders(const QMap<QString, QString> &putHeaders)
+{
+    d->putHeaders.clear();
+    for (QString &name : putHeaders.keys()) {
+        if (name == "Authorization" || name == "Cookie" || name == "Expires") {
+            // newlines must be removed
+            QString value = putHeaders[name];
```

I'm not convinced "removing" newlines is the best option. Also I expect \r isn't accepted either?

Wouldn't failing loudly be more appropriate?

---

I'm not sure, but maybe Qt is already doing this anyway, so we might don't need to remove newlines on our own.